inc/c_util/tarray.h

```diff
@@ -0,0 +1,177 @@
+int TARRAY_ENSURE_CAPACITY(T)(TARRAY(T) tarray, uint32_t capacity)                                                                                              \
+{                                                                                                                                                               \
+    int result;                                                                                                                                                 \
+    if (tarray == NULL)                                                                                                                                         \
+    {                                                                                                                                                           \
+        /*Codes_SRS_TARRAY_02_005: [ If tarray is NULL then TARRAY_ENSURE_CAPACITY(T) shall fail and return a non-zero value. ]*/                               \
+        LogError("invalid argument TARRAY(" MU_TOSTRING(T) ") tarray=%p, uint32_t capacity=%" PRIu32 "",                                                        \
+            tarray, capacity);                                                                                                                                  \
+        result = MU_FAILURE;                                                                                                                                    \
+    }                                                                                                                                                           \
+    else                                                                                                                                                        \
+    {                                                                                                                                                           \
+        /*Codes_SRS_TARRAY_02_006: [ If tarray->capacity is greater than or equal to capacity then TARRAY_ENSURE_CAPACITY(T) shall succeed and return 0. ]*/    \
+        if (capacity <= tarray->capacity)                                                                                                                       \
+        {                                                                                                                                                       \
+            /*we already have the capacity*/                                                                                                                    \
+            result = 0;                                                                                                                                         \
+        }                                                                                                                                                       \
```

Comment:
So, this array never shrinks... Why not have it smart on shrinking too?

Reply:
I guess it is one of those cases "when time comes". Looks like it would be some easy implementation.